File: src/ui/clientForm.ts
import { Client } from '../domains/client';

export class ClientFormState {
  readonly clientId: string;
  readonly branding: string;
  readonly client: Client | null;
  readonly error: string;

  constructor(
    clientId: string = '',
    branding: string = '',
    client: Client | null = null,
    error: string = ''
  ) {
    this.clientId = clientId;
    this.branding = branding;
    this.client = client;
    this.error = error;
  }

  isValid(): boolean {
    return this.clientId.trim() !== '' && this.branding.trim() !== '';
  }
}

export class ClientFormActions {
  updateClientId(state: ClientFormState, clientId: string): ClientFormState {
    return new ClientFormState(
      clientId,
      state.branding,
      state.client,
      ''
    );
  }

  updateBranding(state: ClientFormState, branding: string): ClientFormState {
    return new ClientFormState(
      state.clientId,
      branding,
      state.client,
      ''
    );
  }

  createClient(state: ClientFormState): ClientFormState {
    try {
      if (!state.clientId.trim()) {
        return new ClientFormState(
          state.clientId,
          state.branding,
          null,
          'Client ID cannot be empty'
        );
      }

      const client = new Client({
        id: state.clientId,
        branding: {
          name: state.branding,
          logo: '',
          primaryColor: '#000000'
        }
      });
      return new ClientFormState(
        state.clientId,
        state.branding,
        client,
        ''
      );
    } catch (error) {
      return new ClientFormState(
        state.clientId,
        state.branding,
        null,
        error instanceof Error ? error.message : 'Unknown error'
      );
    }
  }
}

File: python/ui/client_form.py
from dataclasses import dataclass
from typing import Optional
from python.domains.client import Client


@dataclass(frozen=True)
class ClientFormState:
    client_id: str = ""
    branding: str = ""
    client: Optional[Client] = None
    error: str = ""
    
    def is_valid(self) -> bool:
        return self.client_id.strip() != "" and self.branding.strip() != ""


class ClientFormActions:
    def update_client_id(self, state: ClientFormState, client_id: str) -> ClientFormState:
        return ClientFormState(
            client_id=client_id,
            branding=state.branding,
            client=state.client,
            error=""
        )
    
    def update_branding(self, state: ClientFormState, branding: str) -> ClientFormState:
        return ClientFormState(
            client_id=state.client_id,
            branding=branding,
            client=state.client,
            error=""
        )
    
    def create_client(self, state: ClientFormState) -> ClientFormState:
        try:
            if not state.client_id.strip():
                return ClientFormState(
                    client_id=state.client_id,
                    branding=state.branding,
                    client=None,
                    error="Client ID cannot be empty"
                )
            
            client_data = {
                'id': state.client_id,
                'branding': {
                    'name': state.branding,
                    'logo': '',
                    'primary_color': '#000000'
                }
            }
            client = Client(client_data)
            return ClientFormState(
                client_id=state.client_id,
                branding=state.branding,
                client=client,
                error=""
            )
        except Exception as e:
            return ClientFormState(
                client_id=state.client_id,
                branding=state.branding,
                client=None,
                error=str(e)
            )